spec/requests/users/sessions_spec.rb

```diff
@@ -28,6 +46,32 @@
       expect(assigns(:login_token)).to eq(user.login_token)
       expect(response).to render_template(:redirect_from_magic_link)
     end
+
+    it "redirects to sign in when the token doesn't match" do
+      get "/users/confirm_sign_in?login_token=aaaaaaaaaa"
+
+      expect(response).to redirect_to "/users/sign_in"
+    end
+
+    context "when the token has expired" do
+      before { user.update!(login_token_valid_until: 1.hour.ago) }
+
+      it "redirects to sign in when the token has expired" do
+        get "/users/confirm_sign_in?login_token=#{user.login_token}"
+
+        expect(response).to redirect_to "/users/sign_in"
+      end
+    end
+
+    context "when already signed in" do
+      before { sign_in user }
+
+      it "redirects to the dashboard when already signed in" do
+        get "/users/confirm_sign_in?login_token=aaaaaaaaaa"
+
+        expect(response).to redirect_to "/dashboard"
+      end
+    end
```

Comment:
Ditto, context for a single test with duplicated context message.

Reply:
Personally, I prefer this way especially if one is actually setting the context like Ethan did in before { sign_in user }. There are times when people just add unnecessary contexts too but for me this is not one of them.